systemd: 239 -> 242

andir opened the pull request:
###### Motivation for this change

This is a follow-up to #56184.

As explained in https://github.com/NixOS/nixpkgs/pull/56184#issuecomment-490384534 I rebased @Mic92's attempt for systemd version 242 and added the two patches that I wrote last year.

The current head of my systemd fork is at https://github.com/andir/systemd/commits/nixos-v242. I had to add one additional patch to make udev rules work again (https://github.com/andir/systemd/commit/7283141b1bab932e24e2754c261a7e20f4fc3bfa).

While running all the NixOS release.nix tests I discovered a few issues that required changes to our modules. Those are included in this PR.

~~From all the failing tests only the `mysqlReplication` test is seemingly relevant. It requires some refactoring in order to work on v242. The failure log can be seen here: https://gist.github.com/a46e172235ae9066afd758cbf9e86564 It would be great if someone with knowledge in the area could have a look.~~ It was broken before. Nothing we've to take care of.

I am in the process of switching my notebook to this branch to see if anything that we aren't testing is obviously broken.

Feedback from others is highly appreciated.

My private hydra build of `release.nix`: https://hydra.h4ck.space/eval/155 (ipv6 only)

If we like these changes it would be nice if someone with access to https://github.com/nixos/systemd could push my systemd changes to the `nixos-v242` branch.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Mic92 committed:
systemd: 239.20190219 -> 241.20190221

Mic92 committed:
nixos/os-release: add documentation url

andir committed:
systemd: 241.20190221 -> 242

andir committed:
nixos/udev: switch `networking.usePredicatableInterfaceNames` to a kernel param

The udev rules we are shipping no longer work with systemd v242 and were
remove upstream some time ago. It seems like the entire renaming is now
done in C and not in the udev rules.

andir committed:
nixos/tests/radicale: be a bit more permissive when matching logs

With the systemd update to v242 five lines are not longer sufficient to
verify that the storage was verified. In order to reduce future test
failures increasing it to 10 lines sounds like a sane amount.

andir committed:
nixos/networkd: use `no` instead of `none` for `DHCP=` option

systemd has deprecated the use of `none` and recommends using `no`
instead.

andir committed:
nixos/networkd: use the route section for default routes

With systemd v242 using the `Gateway` attribute of the `[Network]`
section will lead to "onlink" routes on all the device that are matched
by the default configuration (typically all devices) causing multiple
default routes (even on localhost).

We can only avoid that - while keeping our default route option - when
we mark the route as explicitly not on link. Only gateways that are
within a subnet of one of the assigned interface addresses will be
installed into the routing table.

andir committed:
nixos/networkd: rename GatewayOnlink to GatewayOnLink

This follows upstreams renaming of the option [1].

[1] https://github.com/systemd/systemd/commit/9cb8c5593443d24c19e40bfd4fc06d672f8c554c

flokli committed:
nixosTests.containers-imperative: add tmpfiles test

(cherry picked from commit 92600a90e248aa27f2aedcce4ad309f987a390df)

andir commented on nixos/modules/services/hardware/udev.nix:
```diff
@@ -282,6 +278,8 @@ in
 
     services.udev.path = [ pkgs.coreutils pkgs.gnused pkgs.gnugrep pkgs.utillinux udev ];
 
+    boot.kernelParams = mkIf (!config.networking.usePredictableInterfaceNames) [ "net.ifnames=0" ];
```
Does this work for all our cases? Previously we did exclude a udev rule. That could have affected containers and similar constructs. Now we only affect "real" systems.

fpletz replied:
tl;dr I don't think this should affect containers or custom network namespace applications.

As I understand it, the interface renaming should only effectively applied to physical interfaces (or if the kernel declares it as predictable) because the physical location is coded into the interface name (see `NamePolicy` in `systemd.link(5)` manpage). When creating virtual netdevs like vlans or bridges a custom name has to be supplied anyway. When moving physical interfaces to containers the interface has already been renamed on the host.

andir replied:
:+1: I created an issue for this (#61629) and also added it to our systemd project board https://github.com/NixOS/nixpkgs/projects/22. (I am trying to populate that more with all the things we come across wile working on this PR).

andir commented on nixos/modules/tasks/network-interfaces-systemd.nix:
```diff
@@ -59,7 +59,14 @@ in
             DHCP = override (dhcpStr cfg.useDHCP);
           } // optionalAttrs (gateway != [ ]) {
-            gateway = override gateway;
+            routes = override [
```
I am not convinced this is the best way to solve this right now.

The issue was that we would add `onlink` routes for all network devices on the system. Introduced in https://github.com/systemd/systemd/commit/4912ab7774ccc1eed07bfbbc6b87e8a23b1f5495 systemd would set the `onlink` attribute for all routes that didn't come with static addresses.

fpletz replied:
It's certainly not ideal but then again adding the default gateway to every interface is FUBAR anyway. This fix is IMHO fine until we finally deprecate `networking.defaultGateway` and move it to the per interface configurations.

andir replied:
:+1: I created an issue for this (#61629) and also added it to our systemd project board https://github.com/NixOS/nixpkgs/projects/22. (I am trying to populate that more with all the things we come across wile working on this PR).

andir commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -14,40 +14,19 @@
   # Also fresh patches should be cherry-picked from that tree to our current one.
   src = fetchFromGitHub {
-    owner = "NixOS";
+    owner = "andir";
```
We should change this to our fork once we know that we want these changes.

Mic92 replied:
Maybe you can also get commit access to our systemd fork.

andir replied:
Moved the source to our repo and adjusted the src attribute.

petabyteboy commented:
I'm testing this now

andir commented:
You might want to use `nixos-rebuild boot`. I lost control of my pid1 after
switching. No idea what we can do about that.

On Sun, 12 May 2019, 11:58 Milan, <notifications@github.com> wrote:

> I'm testing this now
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/61321#issuecomment-491581626>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAE365F4R27JOIPPSMSPWTDPU7S4RANCNFSM4HMJAV2A>
> .
>

petabyteboy commented:
Yup, saw that too late, this is what it looked like:
https://termbin.com/scs1
After a reboot everything seems to work fine until now. I will report here if I find any issues.

flokli commented:
we should probably also check why the failure with https://github.com/NixOS/nixpkgs/pull/56265#issuecomment-491546026 can't be reproduced anymore, and drop the systemd patch referenced there, if it's not effective anymore.

flokli commented:
I figured out why we can't reproduce the error anymore - see explanation in https://github.com/NixOS/nixpkgs/pull/56265#issuecomment-491636865.

@andir, could you incorporate https://github.com/flokli/nixpkgs/commit/92600a90e248aa27f2aedcce4ad309f987a390df into this PR, and drop https://github.com/andir/systemd/commit/51077a9c021431437d426037c2a8281cc21df38c from your systemd branch?

I'll try having a look at how to best fix `nixops send-keys`.

Mic92 commented:
@andir loosing control as in no longer being able to contact via dbus?

petabyteboy commented:
After switching to a new system configuration, systemd-timesyncd started crashing:

```
May 14 11:47:12 tachyon.pbb.lc systemd[21126]: systemd-timesyncd.service: Failed to set up special execution directory in /var/lib: Not a directory
May 14 11:47:12 tachyon.pbb.lc systemd[21126]: systemd-timesyncd.service: Failed at step STATE_DIRECTORY spawning /nix/store/5rnzb2cr3vw536px3p156d1g3arf18v0-systemd-242/lib/systemd/systemd-timesyncd: Not a directory
```

/var/lib exists and is a directory.

andir commented:
> @andir loosing control as in no longer being able to contact via dbus?

Yes, the output looked very similar to https://github.com/NixOS/nixpkgs/pull/61321#issuecomment-491588363.



> Not 100% certain this is related
> After switching to a new system configuration, systemd-timesyncd started crashing:
> 
> ```
> May 14 11:47:12 tachyon.pbb.lc systemd[21126]: systemd-timesyncd.service: Failed to set up special execution directory in /var/lib: Not a directory
> May 14 11:47:12 tachyon.pbb.lc systemd[21126]: systemd-timesyncd.service: Failed at step STATE_DIRECTORY spawning /nix/store/5rnzb2cr3vw536px3p156d1g3arf18v0-systemd-242/lib/systemd/systemd-timesyncd: Not a directory
> ```
> 
> /var/lib exists and is a directory.

I am having the same. It seems like the daemon never came up after switching to the new systemd version. I have some idea what might cause this and will debug this tonight.

petabyteboy commented:
The timesyncd issue has been reported for Arch-based distros too:
https://github.com/systemd/systemd/issues/12131

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/what-are-your-goals-for-19-09/2875/22

flokli commented:
There's some more WIP patches over at https://github.com/flokli/nixpkgs/commits/systemd-v242-fk (not merged in yet here).

Planning to add some more tests for LVM and crypted volumes, plus testing of the `send-keys` functionality, to ensure it doesn't break.

andir commented:
I spent a bit of time on the systemd-timesyncd issue.

Removing the old symlink `/var/lib/systemd/timesync` fixed it for me
locally. It pointed to `../private/systemd/timesync`
(`/var/lib/private/systemd/timesync`) since the unit was previously
running with `DyanmicUser=1`.

The issue here seems to be that even thought the permissions were okay
timesyncd doesn't follow symlinks (anymore?) and thus fails to access
the path that was symlinked.

Upon the next start the directory (/var/lib/systemd/timesync) was
created.  That is definitly not a fix that I'd like to ship to users.

Looking at the changes bit more lead me to the upstream commit 162e0b7
[0] where dynamic users was removed from systemd-timesyncd. Following
that to the related PR 10117 [1] referrenced downstream issues from
Debian and Fedora. The Debian issue seems to be about some confusion
regarding "unknown users" from DBus. Following the Debian issue led to
the bug 9503 [3] where a lengthy discussion and a commit that we've been
shipping in v239 [4]. So I didn't chase that route further.

Looking at how Fedora dealt with the issue brought me to their package
spec [5] file. They check if `/var/lib/systemd/timesync` is a symlink
and in that case they remove it and move the data from
`/var/lib/private/systemd/timesync` to `/var/lib/systemd/timesync`.

At this point I think it might be a safe way to do that in an automated
fashion when switching to a new generation - before starting
systemd-timesyncd.

Any thoughts?


/cc @flokli @petabyteboy

[0] https://github.com/systemd/systemd/commit/162e0b75f9c9f698f94c228c2f9148120f03e9a2
[1] https://github.com/systemd/systemd/pull/10117
[3] https://github.com/systemd/systemd/issues/9503
[4] https://github.com/NixOS/systemd/commit/b179909d5af5318ec2b4270f1acbbc65d033d113
[5] https://src.fedoraproject.org/rpms/systemd/blob/b9896dc2cebc3c4cfda0a6c355183f645e372765/f/systemd.spec#_606

andir commented:
I dropped the commit @flokli mentioned in https://github.com/NixOS/nixpkgs/pull/61321#issuecomment-491637112. Rerunning all the tests :tada:

andir committed:
nixos/systemd: migrate systemd-timesync state when required

Somewhen between systemd v239 and v242 upstream decided to no longer run
a few system services with `DyanmicUser=1` but failed to provide a
migration path for all the state those services left behind.

For the case of systemd-timesync the state has to be moved from
/var/lib/private/systemd/timesync to /var/lib/systemd/timesync if
/var/lib/systemd/timesync is currently a symlink.

We only do this if the stateVersion is still below 19.09 to avoid
starting to have an ever growing activation script for (then) ancient
systemd migrations that are no longer required.

See https://github.com/systemd/systemd/issues/12131 for details about
the missing migration path and related discussion.

c0bw3b commented on nixos/doc/manual/release-notes/rl-1909.xml:
```diff
@@ -153,7 +163,21 @@
+  </listitem>
+  <listitem>
+    <para>
+      With the upgrade to systemd version 242 the <literal>systemd-timesync</literal>
```
```suggestion
      With the upgrade to systemd version 242 the <literal>systemd-timesyncd</literal>
```

andir replied:
Changed. Thanks!

c0bw3b commented on nixos/doc/manual/release-notes/rl-1909.xml:
```diff
@@ -153,7 +163,21 @@
+  <listitem>
+    <para>
+      With the upgrade to systemd version 242 the <literal>systemd-timesync</literal>
+      service is no longer using <literal>DyanmicUser=1</literal>. In order for the
```
```suggestion
      service is no longer using <literal>DynamicUser=yes</literal>. In order for the
```

andir replied:
Changed. Thanks!

c0bw3b commented:
> At this point I think it might be a safe way to do that in an automated fashion when switching to a new generation - before starting systemd-timesyncd.

Agreed. But what happens on rollback or when you manually boot an older gen?

andir commented:
> > At this point I think it might be a safe way to do that in an automated fashion when switching to a new generation - before starting systemd-timesyncd.
> 
> Agreed. But what happens on rollback or when you manually boot an older gen?

As far as I can tell right now that migration would just work. The issue seems to be that a) systemd thinks the StateDirectory is set up and that systemd-timesyncd isn't allowed to follow symlinks anymore and thus fails on newer setups.

On the old setup I expect systemd to migrate the unit to the dynamic user setup or simple start fresh. Starting fresh isn't really tragic for this since as far as I can tell it just stores a timestamp that is used against going backwards in time. It is of course critical if you can't trust your time source. At the point of a rollback there isn't much more we can do. As they might roll back to any channel that has been released up until now.

andir commented:
I will eventually test the rollback on my laptop and then the forward migration to verify my theories. Everyone else is welcome as well :)

andir commented on nixos/doc/manual/release-notes/rl-1909.xml:
```diff
@@ -153,7 +163,21 @@
+    </para>
+    <para>
+      As long as the <literal>system.config.stateVersion</literal> is below
+      <literal>19.09</literal> we will migrate the state folder to it's proper location
```
Done

fpletz committed:
multipath-tools: 0.6.2 -> 0.8.1

andir commented:
@fpletz Can the multipath-tools bump be merged regardless of systemd? It would probably be easier to just merge that into master and only focus on strictly required changes in this branch.

fpletz commented:
It contains a fix for the changed version string in systemd. I will push it to master.

andir committed:
systemd: disable building tests

We are currently not running any tests but building them takes
signitifcant amounts of time since they account to about 40% of all the
compilation targets.

andir committed:
systemd: remove references to $out/lib/systemd/catalog

On aarch64 we "leak" a reference to $out/lib/systemd/catalog in the lib
output. The result of that is a dependency cycle between $out and $lib.
Thus nix (rightfully) marks the build as failed. That reference
originates from an array of strings (catalog_file_dirs) in systemd
(src/src/journal/catalog.{c,h}).  The only consumer (as of v242) of the
symbol is the main function of journalctl.  Still libsystemd.so contains
the VALUE but not the symbol.  Systemd seems to be properly using
function & data sections together with the linker flags to garbage
collect unused sections (-Wl,--gc-sections).  For unknown reasons those
flags do not eliminate the unused string constants, in this case on
aarch64-linux. The hacky way is to just remove the reference after we
finished compiling.  Since it can not be used (there is no symbol to
actually refer to it) there should not be any harm.  It is a bit odd and
I really do not like starting these kind of hacks but there doesn't seem
to be a straight forward way at this point in time.

The reference will be replaced by the same reference the usual nukeRefs
tooling uses.  The standard tooling can not / should not be uesd since
it is a bit too excessive and could potentially do us some (more) harm.

samueldr commented:
About `enable efi on aarch64`. Tested the build, native works (as expected) though looks like the issues with cross-compilation still happen.

Though, through a cursory look I *think* it's a meson misconfiguration for cross. Currently looking more in depth.

samueldr commented:
Turns out the update wasn't needed, as the behaviour is identical. #61847 fixes meson so systemd can find what's required.

The change made to the meson configuration isn't affecting current cross-builds of systemd, it seems that only `systemd-boot` used `cpu_family` from meson.

andir committed:
nixos/test: remove the stateVersion statement from the test-instrumentation

We set stateVersion to `mkDefault 18.03` in
`nixos/modules/testing/test-instrumentation.nix` and in
`modules/installer/cd-dvd/installation-cd-base.nix`.

Accessing the stateVersion in the module system from within the tests
results in the following error:
> The unique option `system.stateVersion' is defined multiple times, in
> `nixpkgs/nixos/modules/installer/cd-dvd/installation-cd-base.nix' and
> `nixpkgs/nixos/modules/testing/test-instrumentation.nix'.

There are other tests that use it as well. Namely the radicale test also
verifies behaviour between state versions is as expected. It switches a
package default value. Others switched on the state directory default.
It seems like having the timesyncd switch as part of every rendered
activationScript might cause this weird error.

Removing this line seems like a reasonable thing to do since we actually
set the default to the very same value in the module system. This line
should have been no-op besides the issue that we've two statements
setting it in this very specific case.

andir commented:
I picked the commits from #61847 into this branch since they seem to make a lot of sense. Cross compiling systemd for aarch64 worked.

Infinisil commented:
(This should probably go to staging)

andir commented:
> (This should probably go to staging)

Agreed, I am keeping this against master so I can compare the test results wit those of hydra. Otherwise we are in the stdenv-rebuild game all the time (and other breaking changes as well).

The final stage of this PR will go into staging.

andir committed:
nixos/misc: warn when someone is using the nixops autoLuks module

The autoLuks module is not really compatible with the updated systemd
version anymore. We started dropping NixOS specific patches that caused
unwanted side effects that we had to work around otherwise.

This change points users towards the relevant PR and spits out a bit of
information on how to deal with the situation.

andir commented:
I think we are reaching the point where we can consider merging (and retargeting to staging once we are happy).

@flokli and I  worked on an issue with the NixOps feature `autoLuks`. Adding the `_netdev` mount option did fix the issues we were observing. I added a commit to this branch that detects NixOps usage of the feature and warns the user. The warning can be disabled with a newly introduced flag. We opened issue #62211 for deprecating the issue. In a call with @grahamc he suggested breaking it on stable so we get some early feedback. I propose backporting my change that introduced the (light) breakage once this reached master.

Additionally I opened a PR (https://github.com/NixOS/nixops/pull/1156) on NixOps that adds some safeguards there. It isn't merged and probably wont be for a while. Removing the feature might be the better way to go.

I'd like to encourage everyone involved in this to revisit the changes. Run a test machine/VM/… to make sure this will be a smooth migration. I can provide a binary cache if you trust me :-)

@petabyteboy did you run into any more troubles?

@Mic92 / @fpletz: Whom shall I contact regarding access to the systemd fork?

petabyteboy commented:
I am using this since a few weeks without any issues.

andir commented:
Moved the systemd source to our systemd repo (https://github.com/NixOS/systemd/commits/nixos-v242)

andir commented:
Dropped the mysql changes that already landed in master & rebased onto latest unstable channel.

flokli commented:
Running this since some days now as a daily driver - except the small issue reported in https://github.com/rycee/home-manager/pull/725, I didn't encounter anything weird. 

I'd suggest to merge this into staging.

fpletz commented:
I've also been running this for a while without issues so far.

andir commented:
Rebased onto staging so we can proceed with merging it.

edolstra commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -207,13 +183,37 @@ in stdenv.mkDerivation rec {
   # systemd builds is the same, then we can switch between them at
   # runtime; otherwise we can't and we need to reboot.
-  passthru.interfaceVersion = 2;
+  passthru.interfaceVersion = 3;
```
What's the reason for bumping `interfaceVersion`? This is really undesirable because it requires every user to reboot.

andir replied:
After re-exec the userspace was no longer able to talk to the daemon. 

See https://github.com/NixOS/nixpkgs/pull/61321#issuecomment-491582147 & https://github.com/NixOS/nixpkgs/pull/61321#issuecomment-491588363. Unfortunately the paste did expire :/

If we come up with a better way to handle these kinds of scenarios that would be great.

Rebooting every once in a while after your pid 1 did a major version change doesn't sound unreasonable to me.

edolstra replied:
Sounds like an upstream systemd bug that should be reported there. AFAIK systemd is supposed to be able to re-exec itself across version changes.

Ma27 commented:
@fpletz @andir Not too experienced with udev and systemd internals, but this seems to break when using a predictably named network interface during a stage-1 boot (in my case to provide SSH access in initrd in order to enter the LUKS passphrase remotely). Until now I've used the `80-net-setup-links.rules` from `nixpkgs` as described in [NixOS discourse](https://discourse.nixos.org/t/how-to-enable-predictable-network-interface-names-in-initrd/979) for a predictably named network interface which is configured like this:

``` nix
{
  boot.kernelParams = [
    "ip=<v4-addr>::<gateway>:255.255.0.0::ens3:none"
  ];
}
```

May I ask what's the recommended way to implement such a feature in NixOS 19.09? AFAICT only udev is started in stage 1, so there's no predicatably named network interface then. I also checked the `80-net-setup-links.rules` from our systemd fork (https://github.com/NixOS/systemd/blob/nixos-v243/rules/80-net-setup-link.rules) which doesn't rename the `eth0` interface in my case (not sure why though, sorry!).

This is why I decided to continue using the old udev rules from nixpkgs (https://github.com/NixOS/nixpkgs/blob/release-19.03/nixos/modules/services/hardware/80-net-setup-link.rules) for now as my network is configured properly in stage-1 then, but I'm not sure if that's a suitable solution :)

Mic92 commented:
I think this update might broke `systemd --user` for me. I am using `lightdm` as a login manager.

andir commented:
On Wed, 2019-09-11 at 03:49 -0700, Jörg Thalheim wrote:
> I think this update might broke systemd --user for me. I am using
> lightdm as a login manager.

What kind of error are you running into? My user session has been
working like before. I am running a bunch of services (through home-
manager) in my user session.

Mic92 commented:
I still see my systemd user session, both the process as well as the dbus session - systemctl currently does not connect to those...

andir commented:
On Wed, 2019-09-11 at 03:56 -0700, Jörg Thalheim wrote:
> I still see my systemd user session, both the process as well as the
> dbus session - systemctl currently does not connect to those...

Ths is indeed odd. I can connect to mine just fine:

```
$ systemctl --user status timers.target 
● timers.target - Timers
   Loaded: loaded (/nix/store/niw0jbw29x0rg85m7z8j5gll16y37g5n-systemd-
242/example/systemd/user/timers.target; static; vendor preset: enabled)
   Active: active since Tue 2019-09-10 00:04:14 CEST; 1 day 12h ago
     Docs: man:systemd.special(7)
```

Can you check what `strace -s 1500 -e connect,recvmsg,sendmsg --
systemctl --user status timers.target` produces?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/predictable-network-interface-names-in-initrd/4055/6

arianvp commented:
I am having an issue with my `systemd --user` too, but slightly different.     It is certainly running:

```
● user@0.service - User Manager for UID 0
   Loaded: loaded (/nix/store/niw0jbw29x0rg85m7z8j5gll16y37g5n-systemd-242/example/systemd/system/user@.service; static; vendor preset: enabled)
  Drop-In: /nix/store/kiiihixa3dcbb947sn4pwq5z90sksijn-system-units/user@.service.d
           └─overrides.conf
   Active: active (running) since Fri 2019-09-13 10:19:56 UTC; 4 days ago
     Docs: man:user@.service(5)
 Main PID: 25811 (systemd)
   Status: "Startup finished in 45ms."
       IP: 0B in, 0B out
    Tasks: 2
   Memory: 3.4M
      CPU: 113ms
   CGroup: /user.slice/user-0.slice/user@0.service
           └─init.scope
             ├─25811 /nix/store/4vw3gb6dk116y38vwgjv3ymq5mfdcfli-systemd-243/lib/systemd/systemd --user
             └─25812 (sd-pam)

Sep 13 10:19:56 arianvp.me systemd[1]: Starting User Manager for UID 0...
Sep 13 10:19:56 arianvp.me systemd[25811]: pam_unix(systemd-user:session): session opened for user root by (uid=0)
Sep 13 10:19:56 arianvp.me systemd[25811]: Reached target Paths.
Sep 13 10:19:56 arianvp.me systemd[25811]: Reached target Sockets.
Sep 13 10:19:56 arianvp.me systemd[25811]: Reached target Timers.
Sep 13 10:19:56 arianvp.me systemd[25811]: Reached target Basic System.
Sep 13 10:19:56 arianvp.me systemd[1]: Started User Manager for UID 0.
Sep 13 10:19:56 arianvp.me systemd[25811]: Reached target Main User Target.
Sep 13 10:19:56 arianvp.me systemd[25811]: Startup finished in 45ms.
```

but  it cannot connect to the bus:
```
[root@arianvp:~]# systemctl status --user
Failed to connect to bus: No such file or directory

```
and if I look at `systemctl status` it seems there is no `dbus.service` running for the `systemd --user` instance:
strace:

```
(...snip..)
openat(AT_FDCWD, "/run/current-system/sw/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0444, st_size=128690032, ...}) = 0
mmap(NULL, 128690032, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f5331506000
close(3)                                = 0
openat(AT_FDCWD, "/proc/self/stat", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0444, st_size=0, ...}) = 0
read(3, "24430 (systemctl) R 24427 24427 "..., 1024) = 320
read(3, "", 1024)                       = 0
close(3)                                = 0
prlimit64(0, RLIMIT_NOFILE, {rlim_cur=512*1024, rlim_max=512*1024}, NULL) = 0
rt_sigaction(SIGBUS, {sa_handler=0x7f53397a9920, sa_mask=[], sa_flags=SA_RESTORER|SA_SIGINFO, sa_restorer=0x7f53396ac860}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
ioctl(1, TCGETS, {B38400 opost isig icanon echo ...}) = 0
newfstatat(AT_FDCWD, "/proc/1/root", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
newfstatat(AT_FDCWD, "/", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
getpid()                                = 24430
socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 3
getsockopt(3, SOL_SOCKET, SO_RCVBUF, [212992], [4]) = 0
setsockopt(3, SOL_SOCKET, SO_RCVBUF, [8388608], 4) = 0
getsockopt(3, SOL_SOCKET, SO_SNDBUF, [212992], [4]) = 0
setsockopt(3, SOL_SOCKET, SO_SNDBUF, [8388608], 4) = 0
connect(3, {sa_family=AF_UNIX, sun_path="/run/user/0/systemd/private"}, 30) = -1 ECONNREFUSED (Connection refused)
(..snip...)
connect(3, {sa_family=AF_UNIX, sun_path="/run/user/0/bus"}, 18) = -1 ENOENT (No such file or directory)
close(3)                                = 0
openat(AT_FDCWD, "/nix/store/6yaj6n8l925xxfbcd65gzqx3dz7idrnn-glibc-2.27/share/locale/locale.alias", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0444, st_size=2997, ...}) = 0
read(3, "# Locale name alias data base.\n#"..., 4096) = 2997
read(3, "", 4096)                       = 0

writev(2, [{iov_base="Failed to connect to bus: No suc"..., iov_len=51}, {iov_base="\n", iov_len=1}], 2Failed to connect to bus: No such file or directory
) = 52
exit_group(1)                           = ?
+++ exited with 1 +++
```

arianvp commented:
FYI, rebooting my machine fixed this for me.   It broke when doing `nixos-rebuild switch` between `239 -> 243` .   We should probably tell people to reboot when going from `19.03` to `19.09`

andir commented:
If you were using the normal switching code you would have seen the
incompatibility noted there.

On Thu, 19 Sep 2019, 12:14 Arian van Putten, <notifications@github.com>
wrote:

> FYI, rebooting my machine fixed this for me. It broke when doing nixos-rebuild
> switch between 239 -> 243 . We should probably tell people to reboot when
> going from 19.03 to 19.09
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/61321?email_source=notifications&email_token=AAE365DMJGYUYFUBS3IIEBLQKNGH7A5CNFSM4HMJAV2KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7C6ZBI#issuecomment-533064837>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAE365D7E76KUVBOQHOOO7LQKNGH7ANCNFSM4HMJAV2A>
> .
>

edolstra commented:
You're already required to reboot, but https://github.com/NixOS/nixpkgs/pull/68906 reverts that.
